dipy/align/imaffine.py

```diff
@@ -0,0 +1,715 @@
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        """
+        if self.samples is None:
+            self._update_dense(xopt, False)
+        else:
+            self._update_sparse(xopt, False)
+        return -1 * self.metric_val
+
+    def gradient(self, xopt):
+        r""" Numeric value of the metric's gradient at the given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_dense(xopt, True)
+        return self.metric_grad * (-1)
+
+    def value_and_gradient(self, xopt):
+        r""" Numeric value of the metric and its gradient at given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_sparse(xopt, True)
+        return -1 * self.metric_val, self.metric_grad * (-1)
+
+
+class AffineRegistration(object):
+    def __init__(self,
+                 metric=None,
+                 level_iters=None,
+                 opt_tol=1e-5,
+                 sigmas=None,
+                 factors=None,
+                 method='L-BFGS-B',
+                 ss_sigma_factor=None,
+                 options=None):
+        r""" Initializes an instance of the AffineRegistration class
+
+        Parameters
+        ----------
+        metric : object
+            an instance of a metric
+        level_iters : list
+            the number of iterations at each level of the Gaussian pyramid.
+            level_iters[0] corresponds to the finest level, level_iters[n-1]
+            the coarsest, where n is the length of the list
+        opt_tol : float
+            tolerance parameter for the optimizer
+        sigmas : list of floats
+            custom smoothing parameter to build the scale space (one parameter
+            for each scale)
+        factors : list of floats
+            custom scale factors to build the scale space (one factor for each
+            scale)
+        method : string
+            optimization method to be used
+        ss_sigma_factor : float
+            parameter of the scale-space smoothing kernel. For example, the
+            std. dev. of the kernel will be factor*(2^i) in the isotropic case
+            where i = 0, 1, ..., n_scales is the scale
+        options : None or dict,
```

Comment:
Yeah, sorry, this is ugly, I am not even using it! =(. I think it's better to put it in `options`.

The long explanation is that, at the beginning, I was not using the Optimizer class, but I implemented my own preconditioned conjugate gradient with golden section line search, and then I thought that nobody was going to accept that, given that we already have the optimization tools, so I removed it and used the Optimizer class instead.